pkg/etcd/etcd.go

```diff
@@ -963,6 +956,98 @@ func (e *ETCD) listSnapshots(ctx context.Context, snapshotDir string) ([]snapsho
 	return snapshots, nil
 }
 
+// initS3IfNil initializes the S3 client
+// if it hasn't yet been initialized.
+func (e *ETCD) initS3IfNil(ctx context.Context) error {
+	if e.s3 == nil {
+		s3, err := newS3(ctx, e.config)
+		if err != nil {
+			return err
+		}
+		e.s3 = s3
+	}
+
+	return nil
+}
+
+// deleteSnapshots removes the given snapshots from
+// either local storage or S3.
+func (e *ETCD) DeleteSnapshots(ctx context.Context, snapshots []string) error {
+	snapshotDir, err := snapshotDir(e.config)
+	if err != nil {
+		return errors.Wrap(err, "failed to get the snapshot dir")
+	}
+
+	if e.config.EtcdS3 {
+		logrus.Info("Removing the given etcd snapshot(s) from S3")
+		logrus.Debugf("Removing the given etcd snapshot from S3: %v", snapshots)
+
+		if e.initS3IfNil(ctx); err != nil {
+			return err
+		}
+
+		objectsCh := make(chan minio.ObjectInfo)
+
+		toCtx, cancel := context.WithTimeout(ctx, defaultS3OpTimeout)
+		defer cancel()
+
+		go func(ctx context.Context) {
+			defer close(objectsCh)
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					opts := minio.ListObjectsOptions{
+						Recursive: true,
+					}
+
+					for obj := range e.s3.client.ListObjects(ctx, e.config.EtcdS3BucketName, opts) {
+						if obj.Err != nil {
+							logrus.Error(obj.Err)
+							continue
+						}
+
+						// iterate through the given snapshots and only
+						// add them to the channel for remove if they're
+						// actually found from the bucket listing.
+						for _, snapshot := range snapshots {
+							if snapshot == obj.Key {
+								objectsCh <- obj
+							}
+						}
+					}
+					return
+				}
+			}
+		}(toCtx)
```

Comment:
This doesn't work as intended, I believe. See my comment here: https://github.com/k3s-io/k3s/pull/3277#discussion_r626980277